### PR TITLE
YONK-829: Created logo/theme against organisation for mobile apps to use

### PR DESCRIPTION
**Pull request by aamishbaloch**
Changes include support of logo/theme against organisation for mobile apps to use.

### CODE REVIEW

**Comment by aamishbaloch** [comment]
@ziafazal Can you review this?

File: mobileapps/views.py [hunk 1]
```diff
@@ -612,3 +622,238 @@ def _create_notification_message(app_id, payload):
...
+    queryset = Theme.objects.all()
+
+
+    def put(self, request, pk, organization_id):
+        data = request.data.copy()
+        data["organization_id"] = organization_id
+
+        theme = Theme.objects.get(pk=pk)
+
+        theme_serializer = ThemeSerializer(theme, data=data)
```
Comment (aamishbaloch) [inline mobileapps/views.py]: @ziafazal Below this line, code is duplicate of the POST view. I am thinking of moving this as a static method against model, a utility method in a separate utils file or make a separate API call of uploading/updating the logo image only in both (POST and PUT) cases. Need your suggestions on these above mentioned options.

File: mobileapps/serializers.py [hunk 2]
```diff
@@ -27,3 +29,22 @@ def create(self, validated_data):
...
+    class Meta:
+        model = Theme
+
+    def get_logo_image(self, theme):
+        return get_logo_image_urls_by_organization_name(
+            theme.organization.name,
+            theme.logo_image_uploaded_at,
+        )
+
+    def create(self, validated_data):
```
Comment (ziafazal) [inline mobileapps/serializers.py]: seems like unnecessary override

File: mobileapps/views.py [hunk 3]
```diff
@@ -612,3 +622,238 @@ def _create_notification_message(app_id, payload):
...
+        theme_serializer = ThemeSerializer(data=data)
+        if theme_serializer.is_valid(raise_exception=True):
+            theme = theme_serializer.save()
+
+        organization = Organization.objects.get(pk=organization_id)
+
+        # validate request:
+        # verify that the user's
+        # ensure any file was sent
+        if 'file' not in request.FILES:
```
Comment (ziafazal) [inline mobileapps/views.py]: This should be move up to perform validation before performing any transactions.

File: mobileapps/views.py [hunk 4]
```diff
@@ -612,3 +622,238 @@ def _create_notification_message(app_id, payload):
...
+            return Response(
+                {
+                    "developer_message": u"No file provided for logo image",
```
Comment (ziafazal) [inline mobileapps/views.py]: can we send only `message` in response.

File: mobileapps/views.py (hunk 1 above)
Comment (ziafazal) [reply mobileapps/views.py]: @aamishbaloch we can create a method i.e `_save_logo` in views.py which should handle the logic around saving log image and then use that method in both POST and PUT handlers.

File: mobileapps/views.py [hunk 5]
```diff
@@ -612,3 +622,238 @@ def _create_notification_message(app_id, payload):
...
+        theme_serializer = ThemeSerializer(theme, data=data)
+        if theme_serializer.is_valid(raise_exception=True):
+            theme = theme_serializer.save()
+
+        organization = Organization.objects.get(pk=organization_id)
+
+        # validate request:
+        # verify that the user's
+        # ensure any file was sent
+        if 'file' not in request.FILES:
```
Comment (ziafazal) [inline mobileapps/views.py]: Move this up before any transactions are performed.

**Review by ziafazal — CHANGES_REQUESTED** [review]
@aamishbaloch completed first round of review.

**Comment by aamishbaloch** [comment]
@ziafazal Done with the changes and unit tests added. Can you review this?

File: mobileapps/image_helpers.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+def get_logo_image_names(organization_name):
+    """
+    Returns a dict containing the filenames for a complete set of organization logo
+    images, keyed by pixel size.
+    """
+    name = _make_logo_image_name(organization_name)
+    return {size: _get_logo_image_filename(name, size) for size in _ORGANIZATION_LOGO_IMAGE_SIZES_MAP}
+
+
+def _get_default_profile_image_urls():
```
Comment (ziafazal) [inline mobileapps/image_helpers.py]: can we rename this to `_get_default_profile_image_urls`?

File: mobileapps/tests.py [hunk 7]
```diff
@@ -5,24 +5,28 @@
...
 from student.tests.factories import UserFactory
 from xmodule.modulestore.tests.django_utils import (
     ModuleStoreTestCase,
     TEST_DATA_SPLIT_MODULESTORE
 )
+from mobileapps.models import MobileApp, NotificationProvider, Theme
+
 
-from mobileapps.models import MobileApp, NotificationProvider
+TEST_PROFILE_IMAGE_UPLOAD_DT = datetime.datetime(2002, 1, 9, 15, 43, tzinfo=UTC)
```
Comment (ziafazal) [inline mobileapps/tests.py]: Can we rename this to `TEST_LOGO_IMAGE_UPLOAD_DT `?

File: mobileapps/tests.py [hunk 8]
```diff
@@ -859,3 +863,265 @@ def test_mobileapp_organization_users_notifications(self, mock_ua_push_api):
...
+            'name': 'Test Theme',
+            'active': True,
+            'organization': self.organization2.id,
+            'file': file_image,
+        }
+
+        self.do_post_multipart(reverse(
+            'mobileapps-organization-themes', kwargs={'organization_id': self.organization2.id}), data,
+        )
+
```
Comment (ziafazal) [inline mobileapps/tests.py]: We should check 201 response code here.

File: mobileapps/urls.py [hunk 9]
```diff
@@ -18,4 +18,10 @@
     url(r'^(?P<mobile_app_id>[0-9]+)/organization/(?P<organization_id>[0-9]+)/notification$',
         mobile_views.MobileAppOrganizationAllUsersNotifications.as_view(),
         name='mobileapps-organization-all-users-notifications'),
+    url(r'^organization/(?P<organization_id>[0-9]+)/themes$',
+        mobile_views.OrganizationThemeView.as_view(),
+        name='mobileapps-organization-themes'),
+    url(r'^organization/(?P<organization_id>[0-9]+)/themes/(?P<pk>[0-9]+)$',
```
Comment (ziafazal) [inline mobileapps/urls.py]: we should rename id param from `pk` to `theme_id`

File: mobileapps/views.py [hunk 10]
```diff
@@ -12,17 +17,47 @@
...
+    MobileRetrieveUpdateDestroyAPIView,
 )
 from edx_solutions_api_integration.users.serializers import SimpleUserSerializer
 from edx_solutions_api_integration.utils import get_ids_from_list_param
 from edx_solutions_organizations.models import Organization
 from edx_solutions_organizations.serializers import BasicOrganizationSerializer
+from image_helpers import get_logo_image_names, set_has_logo_image, create_logo_images
+from openedx.core.djangoapps.profile_images.exceptions import ImageValidationError
+from openedx.core.djangoapps.profile_images.images import (
+    IMAGE_TYPES, validate_uploaded_image, remove_profile_images)
```
Comment (ziafazal) [inline mobileapps/views.py]: What is the usage of `remove_profile_images `?

File: mobileapps/views.py [hunk 11]
```diff
@@ -12,17 +17,47 @@
...
+from mobileapps.models import MobileApp, NotificationProvider, Theme
+from mobileapps.serializers import MobileAppSerializer, NotificationProviderSerializer, ThemeSerializer
 from mobileapps.tasks import publish_mobile_apps_notifications_task
 
 
+def _save_logo(request, organization):
+    # validate request:
+    # verify that the user's
+    # ensure any file was sent
+    if 'file' not in request.FILES:
```
Comment (ziafazal) [inline mobileapps/views.py]: we should rename logo image field to `logo_image` instead of `file`.

File: mobileapps/views.py [hunk 12]
```diff
@@ -612,3 +647,185 @@ def _create_notification_message(app_id, payload):
...
+        return Theme.objects.filter(active=True, organization_id=self.kwargs['organization_id'])
+
+    def post(self, request, organization_id):
+        """
+        POST method inactive the existing active theme and creates and new active one.
+        """
+        data = request.data.copy()
+        data["organization_id"] = organization_id
+
+        if 'file' not in request.FILES:
```
Comment (ziafazal) [inline mobileapps/views.py]: we should rename logo image field to `logo_image` instead of `file`.

**Review by ziafazal — CHANGES_REQUESTED** [review]
@aamishbaloch completed another round of review. Please see my suggestions.

**Comment by aamishbaloch** [comment]
@ziafazal Changes done. Can you review this?

File: mobileapps/image_helpers.py [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+    name = _make_logo_image_name(organization_name)
+    return {size: _get_logo_image_filename(name, size) for size in _ORGANIZATION_LOGO_IMAGE_SIZES_MAP}
+
+
+def _get_default_logo_image_urls():
+    """
+    Returns a dict {size:url} for a complete set of default organization logo images,
+    used as a placeholder when there are no user-submitted organization logo images.
+
+    TODO The result of this function should be memoized, but not in tests.
```
Comment (ziafazal) [inline mobileapps/image_helpers.py]: nit #TODO

File: mobileapps/views.py [hunk 14]
```diff
@@ -612,3 +646,186 @@ def _create_notification_message(app_id, payload):
...
+            - image_url_small
+            - image_url_xsmall
+            - image_url_medium
+        * name: Name of the theme (Optional)
+        * active: theme is active or not (Optional)
+        * organization_id: organization id to which theme is related
+
+        **POST**
+
+        If the request is successful, the request returns an HTTP 204 response.
```
Comment (ziafazal) [inline mobileapps/views.py]: typo 201 response

File: mobileapps/views.py [hunk 15]
```diff
@@ -612,3 +646,186 @@ def _create_notification_message(app_id, payload):
...
+            - image_url_small
+            - image_url_xsmall
+            - image_url_medium
+        * name: Name of the theme (Optional)
+        * active: theme is active or not (Optional)
+        * organization_id: organization id to which theme is related
+
+        **PATCH**
+
+        If the request is successful, the request returns an HTTP 204 response.
```
Comment (ziafazal) [inline mobileapps/views.py]: I think a successful patch should return HTTP 200?

File: mobileapps/views.py [hunk 16]
```diff
@@ -612,3 +646,195 @@ def _create_notification_message(app_id, payload):
...
+        theme = Theme.objects.get(pk=theme_id)
+        theme_serializer = ThemeSerializer(theme, data=data)
+        if theme_serializer.is_valid(raise_exception=True):
+            theme = theme_serializer.save()
+
+        if 'logo_image' in request.FILES:
+            organization = Organization.objects.get(pk=organization_id)
+            is_logo_saved, response_dict = _save_logo(request, organization)
+
+        return Response(status=status.HTTP_204_NO_CONTENT)
```
Comment (ziafazal) [inline mobileapps/views.py]: This should return 200 http status code.

File: mobileapps/views.py [hunk 17]
```diff
@@ -612,3 +646,195 @@ def _create_notification_message(app_id, payload):
...
+
+    def delete(self, request, theme_id, organization_id):
+        theme = Theme.objects.get(pk=theme_id)
```
Comment (ziafazal) [inline mobileapps/views.py]: can we use a single statement like this here?
```
Theme.objects.filter(pk=theme_id).update(active=None)
```

**Review by ziafazal — CHANGES_REQUESTED** [review]
@aamishbaloch a few minor things if you can address those it would be great.

**Review by ziafazal — APPROVED** [review]
👍

**Pull request merged**